plugins/woocommerce/includes/wc-order-functions.php

```diff
@@ -162,10 +162,27 @@ function wc_get_is_pending_statuses() {
  * @return string
  */
 function wc_get_order_status_name( $status ) {
-	$statuses = wc_get_order_statuses();
-	$status   = OrderUtil::remove_status_prefix( $status );
+	// "Special statuses": these are in common usage across WooCommerce, but are not normally returned by
+	// wc_get_order_statuses(). We add them only to avoid unnecessary doing-it-wrong errors.
+	$special_statuses = array(
+		'wc-' . OrderStatus::AUTO_DRAFT => OrderStatus::AUTO_DRAFT,
+		'wc-' . OrderStatus::TRASH      => OrderStatus::TRASH,
+	);
+
+	// Merge order is important. If the special statuses are ever returned by wc_get_order_statuses(), those definitions
+	// should take priority.
+	$statuses   = array_merge( $special_statuses, wc_get_order_statuses() );
+	$unprefixed = OrderUtil::remove_status_prefix( (string) $status );
+
+	if ( ! is_string( $status ) || ! isset( $statuses[ 'wc-' . $unprefixed ] ) ) {
```

Comment:
If I understood it correctly, this condition would allow any status type as long as it is a string. Do we want to be more rigid about the status types here and allow only pre-defined types?

Reply:
> This condition would allow any status type as long as it is a string.

It's not some much allowing or disallowing, it's just that we will warn via `wc_doing_it_wrong()` if the `$status` doesn't meet our expectations:

- `(float) 3.14` will lead to a warning because it isn't a string.
- `(string) 'Foobarbaz'` will lead to a warning because, at least under default conditions, that isn't a valid status returned by `wc_get_order_statuses()`.

Reply:
> Do we want to be more rigid

So I don't know that we *can* become more rigid ... or not without introducing back-compat problems. 

Possibly this is answering the same thing you are asking [re one of the new assertions](https://github.com/woocommerce/woocommerce/pull/52797/files#r1877920701): if in invalid status is passed in, but it is nonetheless a string, then my feeling is we should still honour the existing contract (which we do, but we also raise awareness of what might be a potential problem by raising a *doing-it-wrong).*

Let me know if I'm misunderstanding you.